src/CoreShop/Bundle/VariantBundle/Service/VariantGeneratorService.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * CoreShop
+ *
+ * This source file is available under two different licenses:
+ *  - GNU General Public License version 3 (GPLv3)
+ *  - CoreShop Commercial License (CCL)
+ * Full copyright and license information is available in
+ * LICENSE.md which is distributed with this source code.
+ *
+ * @copyright  Copyright (c) CoreShop GmbH (https://www.coreshop.org)
+ * @license    https://www.coreshop.org/license     GPLv3 and CCL
+ *
+ */
+
+namespace CoreShop\Bundle\VariantBundle\Service;
+
+
+use CoreShop\Component\Variant\Model\AttributeInterface;
+use CoreShop\Component\Variant\Model\ProductVariantAwareInterface;
+use Pimcore\Model\DataObject;
+use Pimcore\Model\DataObject\AbstractObject;
+use Pimcore\Tool;
+
+class VariantGeneratorService implements VariantGeneratorServiceInterface
+{
+    public function generateVariants(array $combinations, ProductVariantAwareInterface $product): array
+    {
+        $variants = [];
+        foreach($combinations as $attributeIds) {
+            $variant = $this->generateVariant($attributeIds, $product);
+
+            if($variant) {
+                $variants[] = $variant;
+            }
+        }
+
+        return $variants;
+    }
+
+    public function generateVariant(array $attributeIds, ProductVariantAwareInterface $product): ?ProductVariantAwareInterface
+    {
+        if (count($product->getAllowedAttributeGroups()) !== count($attributeIds)) {
+            return null;
+        }
+
+        $class = get_class($product);
```

Comment:
should we check for Pimcore Concrete here and use `$product::getListing()` here?

Reply:
@dpfaffenbauer absolutely